crates/ra_hir/src/ty/infer.rs

```diff
@@ -970,23 +985,160 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
+                ty_app!(TypeCtor::RawPtr(Mutability::Shared)),
+                ty_app!(TypeCtor::RawPtr(Mutability::Mut)),
+            )
+            | (
+                ty_app!(TypeCtor::Ref(Mutability::Shared)),
+                ty_app!(TypeCtor::Ref(Mutability::Mut)),
+            ) => return false,
+
+            // `{function_type}` -> `fn()`
+            (ty_app!(TypeCtor::FnDef(_)), ty_app!(TypeCtor::FnPtr { .. })) => {
+                match from_ty.callable_sig(self.db) {
+                    None => return false,
+                    Some(sig) => {
+                        let num_args = sig.params_and_return.len() as u16 - 1;
+                        from_ty =
+                            Ty::apply(TypeCtor::FnPtr { num_args }, Substs(sig.params_and_return));
+                    }
+                }
+            }
+
+            // Trivial cases, this should go after `never` check to
+            // avoid infer result type to be never
+            _ => {
+                if self.unify_inner_trivial(&from_ty, &to_ty) {
+                    return true;
+                }
+            }
+        }
+
+        // Try coerce or unify
+        match (&from_ty, &to_ty) {
+            // FIXME: Solve `FromTy: CoerceUnsized<ToTy>` instead of listing common impls here.
+            (ty_app!(TypeCtor::Ref(_), st1), ty_app!(TypeCtor::Ref(_), st2))
+            | (ty_app!(TypeCtor::RawPtr(_), st1), ty_app!(TypeCtor::RawPtr(_), st2)) => {
+                match self.try_coerce_unsized(&st1[0], &st2[0], 0) {
+                    Some(ret) => return ret,
+                    None => {}
+                }
+            }
+            _ => {}
+        }
+
+        // Auto Deref if cannot coerce
+        match (&from_ty, &to_ty) {
+            (ty_app!(TypeCtor::Ref(_), st1), ty_app!(TypeCtor::Ref(_), st2)) => {
+                self.unify_autoderef_behind_ref(&st1[0], &st2[0])
+            }
+
+            // Normal unify
+            (ty_app!(ctor1, st1), ty_app!(ctor2, st2)) => {
+                if ctor1 != ctor2 {
+                    false
+                } else {
+                    self.unify_substs(st1, st2, 0)
+                }
+            }
+            _ => false,
         }
     }
 
+    /// Coerce a type to a DST if `FromTy: Unsize<ToTy>`
+    ///
+    /// See: `https://doc.rust-lang.org/nightly/std/marker/trait.Unsize.html`
+    fn try_coerce_unsized(&mut self, from_ty: &Ty, to_ty: &Ty, depth: usize) -> Option<bool> {
+        if depth > 1000 {
+            panic!("Infinite recursion in coercion");
+        }
+
+        // FIXME: Correctly handle
+        match (&from_ty, &to_ty) {
+            // `[T; N]` -> `[T]`
+            (ty_app!(TypeCtor::Array, st1), ty_app!(TypeCtor::Slice, st2)) => {
+                Some(self.unify(&st1[0], &st2[0]))
+            }
+
+            // `T` -> `dyn Trait` when `T: Trait`
+            (_, Ty::Dyn(_)) => {
+                // Always success since we don't need to check predicates here.
```

Comment:
it's not immediately clear to me why not, could you explain? I would expect us to add an obligation here.

Reply:
Oh, it _should_ check trait predicates. I just don't want it to break other types related to it in following codes, since this could not unify anything.

If correctness if more important, then we should check trait predicates here.

Reply:
I'm not sure what you think it could break.

It's not so much about correctness (since we don't report any failures in obligations anyway), but it could actually be necessary for inference in some cases. Imagine we're coercing to some `dyn Trait<T>`, where we don't know the `T` yet; in that case, solving the obligation can actually tell us the `T`.

I don't think you need to add this in this PR though; but the comment sounds like we're intentionally not doing it.